File: classes/question.py
from classes.country import Country

class Question(Country):
    def __init__(self, id, name, capital, continent, question, answer_key):
        super().__init__(id, name, capital, continent)
        self.__question = question
        self.__answer_key = answer_key
        self.__answer = None
            
    @property
    def question(self):
        return self.__question

    @question.setter
    def question(self, question):
        if isinstance(question, str):
            self.__question = question
        else:
            raise ValueError("question must be a string")
        
    @property
    def answer_key(self):
        return self.__answer_key

    @answer_key.setter
    def correct_answer(self, answer):
        if isinstance(answer, str):
            self.__answer_key = answer
        else:
            raise ValueError("answer must be a string")

    @property
    def answer(self):
        return self.__answer

    @answer.setter
    def answer(self, answer):
        if isinstance(answer, str):
            self.__answer = answer.strip()
        else:
            raise ValueError("answer must be a string")

    def add_option(self, option):
        """Add an option to the question."""
        if isinstance(option, str) and option.strip():
            self.__options.append(option)
        else:
            raise ValueError("Option must be a non-empty string")

    def __str__(self):
        base_info = super().__str__()
        return f"{base_info}, Question: {self.__desc}, Answer Key: {self.__answer_key}"
    
    
    def to_dict(self):
        return {
            "id": self.id,
            'name': self.name,
            'capital': self.capital,
            'continent': self.continent,
            'question': self.__question,
            'answer_key': self.__answer_key
        }

File: java/classes/Question.java
package classes;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.Random;
import java.util.Set;

public class Question extends Country {
    private String question;
    private String answerKey;
    private ArrayList<String> options;
    private String answer;

    public Question(int id, String name, String capital, String question, String answerKey) {
        super(id, name, capital);
        this.question = question;
        this.answerKey = answerKey;
        this.options = new ArrayList<>();
        this.answer = null;
    }

    public String getQuestion() {
        return question;
    }

    public void setQuestion(String question) {
        this.question = question;
    }

    public String getAnswerKey() {
        return answerKey;
    }

    public void setAnswerKey(String answerKey) {
        this.answerKey = answerKey;
    }

    public ArrayList<String> getOptions() {
        return options;
    }

    public void setOptions(ArrayList<String> options) {
        if (options != null) {
            this.options = options;
        } else {
            throw new IllegalArgumentException("Options must be a list");
        }
    }

    public String getAnswer() {
        return answer;
    }

    public void setAnswer(String answer) {
        this.answer = answer != null ? answer.trim() : null;
    }

    @Override
    public String toString() {
        return super.toString() + ", Question: " + question + ", Answer Key: " + answerKey;
    }

    public void addOptions(ArrayList<Country> countries) {
        Set<String> optionsSet = new HashSet<>();
        optionsSet.add(this.getCapital());

        Random random = new Random();
        while (optionsSet.size() < 4) {
            Country randomCountry = countries.get(random.nextInt(countries.size()));
            String randomOption = randomCountry.getCapital();
            optionsSet.add(randomOption);
        }

        this.options = new ArrayList<>(optionsSet);
        Collections.shuffle(this.options);
    }
}